flow/connectors/mongo/mongo.go

```diff
@@ -229,46 +231,89 @@ func (c *MongoConnector) PullRecords(
 
 	changeStream, err := c.client.Watch(ctx, defaultPipeline(), changeStreamOpts)
 	if err != nil {
-		var cmdErr mongo.CommandError
-		// ChangeStreamHistoryLost is basically slot invalidation
-		if errors.As(err, &cmdErr) && cmdErr.Code == 286 {
-			return errors.New("change stream history lost")
-		}
-		return err
+		return fmt.Errorf("failed to create change stream: %w", err)
 	}
 	defer changeStream.Close(ctx)
-	c.logger.Info("ChangeStream started for mirror " + req.FlowJobName)
 
 	var recordCount uint32
 	defer func() {
 		if recordCount == 0 {
 			req.RecordStream.SignalAsEmpty()
 		}
-		c.logger.Info(fmt.Sprintf("[finished] PullRecords streamed %d records", recordCount))
+		c.logger.Info("[mongo] PullRecords finished streaming", slog.Uint64("records", uint64(recordCount)))
 	}()
-	// before first record, we wait indefinitely so give it ctx
-	// after first record, we wait for idle timeout
-	getCtx := ctx
-	var cancelTimeout context.CancelFunc
+	// before the first record arrives, we wait for up to an hour before resetting context timeout
+	// after the first record arrives, we switch to configured idleTimeout
+	timeoutCtx, cancelTimeout := context.WithTimeout(ctx, time.Hour)
+	//nolint:gocritic // cancelTimeout is rebound, do not defer cancelTimeout()
 	defer func() {
-		if cancelTimeout != nil {
-			cancelTimeout()
-		}
+		cancelTimeout()
 	}()
+
+	checkpoint := func() {
+		if resumeToken := changeStream.ResumeToken(); resumeToken != nil {
+			resumeTokenText := base64.StdEncoding.EncodeToString(resumeToken)
+			req.RecordStream.UpdateLatestCheckpointText(resumeTokenText)
+		} else {
+			c.logger.Warn("change stream document does not contain a resume token")
+		}
+	}
+
 	addRecord := func(ctx context.Context, record model.Record[model.RecordItems]) error {
 		recordCount += 1
 		if err := req.RecordStream.AddRecord(ctx, record); err != nil {
 			return err
 		}
 		if recordCount == 1 {
 			req.RecordStream.SignalAsNotEmpty()
-			// after the first record, we switch to a timeout context
-			getCtx, cancelTimeout = context.WithTimeout(ctx, req.IdleTimeout)
+			timeoutCtx, cancelTimeout = context.WithTimeout(ctx, req.IdleTimeout)
 		}
 		return nil
 	}
 
-	for recordCount < req.MaxBatchSize && changeStream.Next(getCtx) {
+	recreateChangeStream := func() error {
+		if err := changeStream.Close(timeoutCtx); err != nil {
```

Comment:
seems odd passing timeoutCtx when we're calling this after hitting DeadlineExceeded already

Reply:
ah,  should just be `ctx` here.